From 6837bcd1a698c1b5a361af4ecaab4877bbc80dc0 Mon Sep 17 00:00:00 2001
From: Jason Lowe <jlowe@apache.org>
Date: Thu, 5 Feb 2015 16:40:09 +0000
Subject: [PATCH 1134/2748] MAPREDUCE-6059. Speed up history server startup
 time (Siqi Li via aw) (cherry picked from commit
 fd57ab2002f97dcc83d455a5e0c770c8efde77a4)

Conflicts:

	hadoop-mapreduce-project/CHANGES.txt

(cherry picked from commit 2dc4af2b9379e5257dda4ea5fe551dc8e4af5226)

Change-Id: Ibc5ab1c5534f41e0f5a1009aed34f857a3f3ec21
---
 .../hadoop/mapreduce/v2/hs/HistoryFileManager.java |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java
index c407135..79d7282 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java
@@ -294,6 +294,10 @@ public void delete(HistoryFileInfo fileInfo) {
     public HistoryFileInfo get(JobId jobId) {
       return cache.get(jobId);
     }
+
+    public boolean isFull() {
+      return cache.size() >= maxSize;
+    }
   }
 
   /**
@@ -702,6 +706,10 @@ void initExisting() throws IOException {
     for (FileStatus fs : timestampedDirList) {
       // TODO Could verify the correct format for these directories.
       addDirectoryToSerialNumberIndex(fs.getPath());
+    }
+    for (int i= timestampedDirList.size() - 1;
+        i >= 0 && !jobListCache.isFull(); i--) {
+      FileStatus fs = timestampedDirList.get(i); 
       addDirectoryToJobListCache(fs.getPath());
     }
   }
-- 
1.7.9.5

